keps/sig-apps/0024-ttl-after-finish.md

```diff
@@ -0,0 +1,290 @@
+     the Job after a computed amount of time when it will expire.
+1. Delete the Job if passing the sanity checks. 
+
+#### Finished Pods
+
+When a Pod is created or updated:
+1. Check its `.status.phase` to see if it has finished (`Succeeded` or `Failed`).
+   If it hasn't finished, do nothing. 
+1. Otherwise, if the Pod has finished, check if Pod's
+   `.spec.ttlSecondsAfterFinished` field is set. Do nothing if the TTL field is
+   not set. 
+1. Otherwise, if the TTL field is set, check if the TTL has expired, i.e.
+   `.spec.ttlSecondsAfterFinished` + the time when the Pod finishes (max of all
+   of its containers termination time
+   `.containerStatuses.state.terminated.finishedAt`) > now. 
+1. If the TTL hasn't expired, delay re-enqueuing the Pod after a computed amount
+   of time when it will expire. The computed time period is:
+   (`.spec.ttlSecondsAfterFinished` + the time when the Pod finishes - now).
+1. If the TTL has expired, `GET` the Pod from API server to do final sanity
+   checks before deleting it.
+1. Check if the freshly got Pod's TTL has expired. This field may be updated
+   before TTL controller observes the new value in its local cache.
+   * If it hasn't expired, it is not safe to delete the Pod. Delay re-enqueue
+     the Pod after a computed amount of time when it will expire.
+1. Delete the Pod if passing the sanity checks. 
+
+#### Owner References
+
+We have considered making TTL controller leave a Job/Pod around even after its
+TTL expires, if the Job/Pod has any owner specified in its
+`.metadata.ownerReferences`.
+
+We decided not to block deletion on owners, because the purpose of
+`.metadata.ownerReferences` is for cascading deletion, but not for keeping an
+owner's dependents alive. If the Job is owned by a CronJob, the Job can be
+cleaned up based on CronJob's history limit (i.e. the number of dependent Jobs
+to keep), or CronJob can choose not to set history limit but set the TTL of its
+Job template to clean up Jobs after TTL expires instead of based on the history
+limit capacity. 
+
+Therefore, a Job/Pod can be deleted after its TTL expires, even if it still has
+owners. 
+
+Similarly, the TTL won't block deletion from generic garbage collector. This
+means that when a Job's or Pod's owners are gone, generic garbage collector will
+delete it, even if it hasn't finished or its TTL hasn't expired. 
+
+### Risks and Mitigations
+
+Risks:
+* Time skew may cause TTL controller to clean up resource objects at the wrong
+  time.
+
+Mitigations:
+* In Kubernetes, it's required to run NTP on all nodes ([#6159][]) to avoid time
+  skew. We will also document this risk.
+
+[#6159]: https://github.com/kubernetes/kubernetes/issues/6159#issuecomment-93844058
+
+## Graduation Criteria
+
+This can be promoted to beta when it satisfies users' need for cleaning up
+finished resource objects, without regressions.
+
+This will be promoted to GA once it's gone a sufficient amount of time as beta
```

Comment:
[From the previous dicussions on the google doc](https://docs.google.com/document/d/1U6h1DrRJNuQlL2_FYY_FdkQhgtTRn1kEylEOHRoESTc/edit?disco=AAAACB9yb9k):

*Because metadata is used by all resources, but not all resources are finishable (only Pods/Jobs). We want to implement this for Pods/Jobs first. If users like this feature, and more use cases from custom resources, we can move it to metadata. To avoid the overhead, we can either remove this alpha field in spec and add a field in metadata, or generalize by keeping it in the spec of each finishable resource.*

I am not sure I want us promoting something to beta/GA if we know that we will replace it with a generic mechanism. We should flesh out the path here a bit more in that regard.

Reply:
> This will only be promoted to beta after this decision is finalized, i.e. whether we decide to replace it with a generic mechanism or not to do it. Adding this feature as alpha first is essential to help us gather feedback before deciding to generalize it or not.

I agree we should try this as alpha to gather feedback. I was just asking to formalize what you just said as beta graduation criteria in the proposal.

Reply:
Thanks, updated graduation criteria.